File: python/bj/BJ2447.py
import sys

n = int(sys.stdin.readline().rstrip())
result = [[0 for _ in range(n)] for _ in range(n)]

def make_void(n, y, x):
    tri = int(n / 3)
    for i in range(y + tri, y + (tri * 2)):
        for j in range(x + tri, x + (tri * 2)):
            result[i][j] = 2

def do_recursively(n, y, x):
    make_void(n, y, x)
    if n == 3: return
    tri = int(n / 3)
    for i in range(0, n, tri):
        for j in range(0, n, tri):
            if i == tri and j == tri: continue
            do_recursively(tri, y + i, x + j)

do_recursively(n, 0, 0)

for i in range(n):
    for j in range(n):
        if result[i][j] == 0:
            print('*', end='')
        else:
            print(' ', end='')
    print()

File: java/src/baekjoon/BJ2447.java
package baekjoon;

import java.io.*;

public class BJ2447 {

    // 2 means a void; 0 means a star.
    private static int[][] result;

    private static void makeHole(int n, int y, int x) {
        int tri = n / 3;
        for (int i = y + tri; i < y + tri * 2; ++i) {
            for (int j = x + tri; j < x + tri * 2; ++j)
                result[i][j] = 2;
        }
    }
    private static void doRecursively(int n, int y, int x) {
        makeHole(n, y, x);
        if (n == 3) return;

        int tri = n / 3;
        for (int i = 0; i < n; i += tri) {
            for (int j = 0; j < n; j += tri) {
                if (i == tri && j == tri) continue;
                doRecursively(tri, y + i, x + j);
            }
        }
    }

    private static void solve() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int n = Integer.parseInt(br.readLine());
        result = new int[n][n];
        doRecursively(n, 0, 0);

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (result[i][j] == 2)
                    sb.append(' ');
                else
                    sb.append('*');
            }
            sb.append('\n');
        }

        bw.write(sb.toString());
        bw.flush();
    }

    public static void main(String[] args) throws IOException {
        solve();
    }

}
